evm/src/memory/memory_stark.rs

```diff
@@ -0,0 +1,580 @@
+    for idx in 0..num_ops - 1 {
+        let this_context_first_change = context[idx] != context[idx + 1];
+        let this_segment_first_change =
+            segment[idx] != segment[idx + 1] && !this_context_first_change;
+        let this_virtual_first_change = virtuals[idx] != virtuals[idx + 1]
+            && !this_segment_first_change
+            && !this_context_first_change;
+
+        context_first_change.push(F::from_bool(this_context_first_change));
+        segment_first_change.push(F::from_bool(this_segment_first_change));
+        virtual_first_change.push(F::from_bool(this_virtual_first_change));
+    }
+
+    context_first_change.push(F::ZERO);
+    segment_first_change.push(F::ZERO);
+    virtual_first_change.push(F::ZERO);
+
+    (
+        context_first_change,
+        segment_first_change,
+        virtual_first_change,
+    )
+}
+
+pub fn generate_range_check_value<F: RichField>(
+    context: &[F],
+    segment: &[F],
+    virtuals: &[F],
+    timestamp: &[F],
+    context_first_change: &[F],
+    segment_first_change: &[F],
+    virtual_first_change: &[F],
+) -> Vec<F> {
+    let num_ops = context.len();
+    let mut range_check = Vec::new();
+
+    for idx in 0..num_ops - 1 {
+        let this_address_unchanged = F::ONE
+            - context_first_change[idx]
+            - segment_first_change[idx]
+            - virtual_first_change[idx];
+
+        range_check.push(
+            context_first_change[idx] * (context[idx + 1] - context[idx] - F::ONE)
+                + segment_first_change[idx] * (segment[idx + 1] - segment[idx] - F::ONE)
+                + virtual_first_change[idx] * (virtuals[idx + 1] - virtuals[idx] - F::ONE)
+                + this_address_unchanged * (timestamp[idx + 1] - timestamp[idx] - F::ONE),
+        );
+    }
+
+    range_check.push(F::ZERO);
+
+    range_check
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> MemoryStark<F, D> {
+    pub(crate) fn generate_trace_rows(
+        &self,
+        memory_ops: Vec<MemoryOp<F>>,
+    ) -> Vec<[F; NUM_REGISTERS]> {
+        let num_ops = memory_ops.len();
+
+        let mut trace_cols = [(); NUM_REGISTERS].map(|_| vec![F::ZERO; num_ops]);
+        for i in 0..num_ops {
+            let MemoryOp {
+                timestamp,
+                is_read,
+                context,
+                segment,
+                virt,
+                value,
+            } = memory_ops[i];
+            trace_cols[TIMESTAMP][i] = timestamp;
+            trace_cols[IS_READ][i] = is_read;
+            trace_cols[ADDR_CONTEXT][i] = context;
+            trace_cols[ADDR_SEGMENT][i] = segment;
+            trace_cols[ADDR_VIRTUAL][i] = virt;
+            for j in 0..8 {
+                trace_cols[value_limb(j)][i] = value[j];
+            }
+        }
+
+        self.generate_memory(&mut trace_cols);
+
+        let mut trace_rows = vec![[F::ZERO; NUM_REGISTERS]; num_ops];
+        for (i, col) in trace_cols.iter().enumerate() {
```

Comment:
Looks like this is transposing? Seems unnecessary since the next step would be to transpose back to col-major form.

It's not a big issue though, and I wanted to experiment with some other changes to witness generation, namely doing more of it in row-major form to avoid sparse use of memory (for cache efficiency). So I think we can leave it for now and I can make a PR with various generation tweaks later.